### PR TITLE
Reduce memory in classification metrics when `average='micro'`

### LINKED ISSUE #1279: `MulticlassAccuracy` using lot of memory with many classes
## 🐛 Bug

After the refactor of the Accuracy package, I am experiencing memory problems in the classification with many classes. Basically, I have to compute the accuracy over 50K classes (output of a Transformer trained with MLM). I just need the “micro” accuracy, so something like `(preds == target).mean()`. I do not understand how I could use `MulticlassAccuracy` to avoid the instantiation of a tensor of 50K*50K integers, which requires about 18GB of GPU memory.

### To Reproduce

Just use `MulticlassAccuracy` with a large number of classes.

#### Code sample

```python
accu = MulticlassAccuracy(num_classes=50000, average='micro')
preds = torch.tensor([[0, 1, 2], [1, 1, 2]])
target = torch.tensor([[0, 1, 1], [2, 2, 1]])
accu(preds, target)
```

### Expected behavior

With `micro` average, it not necessary to instantiate a matrix of `num_classes^2` to store the predictions over each class. The problematic line is the following `https://github.com/Lightning-AI/metrics/blob/84ef54a281af00fe93b98ab89696e393969654a9/src/torchmetrics/functional/classification/stat_scores.py#L405`

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): pip 0.10.0
- Python & PyTorch Version (e.g., 1.0): 3.9.12
- Any other relevant information such as OS (e.g., Linux): MPS on macOS 12.6 and Ubuntu 20.04 with CUDA 11.6

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1279 
When `average='micro'` in multiclass classification we do not need to form the full `[n_class, n_class]` confmat to do the calculations. This PR introduces a much simpler reduction that scales much better when `n_class` is a large number.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]
LGTM, can we add a simple test-case asserting that this is actually used properly?

**Review by lucadiliello — APPROVED** [review]
Personally tested and worked flawlessly on MPS, reducing memory usage of the metric from 18GB to a few KB.

**Comment by SkafteNicki** [comment]
> LGTM, can we add a simple test-case asserting that this is actually used properly?

Do you want test for memory consumption?
Because tests for correctness is already covered by the tests that are already in place.

**Review by stancld — APPROVED** [review]

**Comment by Borda** [comment]
> Do you want test for memory consumption?

would be great :)
